### PR TITLE
better color theme typing

### PR DESCRIPTION
**Pull request by arose**
Quick attempt at better color theme typing. Can break things in some cases.

### CODE REVIEW

File: src/mol-repr/volume/slice.ts [hunk 1]
```diff
@@ -32,7 +32,7 @@ export async function createImage(ctx: VisualContext, volume: Volume, theme: The
     const isoVal = Volume.IsoValue.toAbsolute(isoValue, volume.grid.stats).absoluteValue;
 
     // TODO more color themes
-    const color = theme.color.color(NullLocation, false);
+    const color = 'color' in theme.color ? theme.color.color(NullLocation, false) : Color(0xffffff);
```
Comment (dsehnal) [inline src/mol-repr/volume/slice.ts:35]:
```suggestion
    const color = theme.color.color?.(NullLocation, false) ?? Color(0xffffff);
```

File: src/mol-theme/color.ts [hunk 2]
```diff
@@ -64,6 +62,26 @@ interface ColorTheme<P extends PD.Params> {
...
+} & ColorThemeShared<P>
+
+type ColorThemeDirect<P extends PD.Params> = {
+    readonly granularity: ColorTypeDirect
+} & ColorThemeShared<P>
+
+type ColorTheme<P extends PD.Params, G extends ColorType = any> =
+    G extends ColorTypeLocation ? ColorThemeLocation<P> :
+        G extends ColorTypeGrid ? ColorThemeGrid<P> :
+            G extends ColorTypeDirect ? ColorThemeDirect<P> : never
```
Comment (dsehnal) [inline src/mol-theme/color.ts:83]:
```suggestion
            G extends ColorTypeDirect ? ColorThemeDirect<P> : ColorTheme<any>
```

And this should prevent any breaking?

File: src/mol-repr/volume/slice.ts (hunk 1 above)
Comment (arose) [reply src/mol-repr/volume/slice.ts:35]: Doesn't work, need to type narrow first

File: src/mol-theme/color.ts (hunk 2 above)
Comment (arose) [reply src/mol-theme/color.ts:83]: I like to keep `never` so we are forced to handle any new granularity types

**Comment by arose** [comment]
need to think some more about it

**Comment by arose** [comment]
Set `ColorTypeLocation` as the default `ColorType` in `ColorTheme`, so you only need to set the `ColorType` when using `ColorTypeDirect` or `ColorTypeGrid`, which is probably hardly used. 

For example `ColorTheme<VolumeValueColorThemeParams, ColorTypeDirect>`

**Pull request merged**